ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,203 @@
+
+module Ouroboros.Network.PeerSelection.LedgerPeers (
+    DomainAddress (..),
+    IP.IP (..),
+    LedgerPeersConsensusInterface (..),
+    RelayAddress (..),
+    PoolStake (..),
+    AckPoolStake (..),
+    runLedgerPeers,
+    TraceLedgerPeers (..),
+    pickPeers,
+    ackPoolStake,
+
+    Socket.PortNumber
+    ) where
+
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Control.Tracer (Tracer, traceWith)
+import qualified Data.IP as IP
+import           Data.List (foldl', nub)
+import           Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NonEmpty
+import           Data.Maybe (isNothing, fromJust)
+import qualified Data.Map.Strict as Map
+import           Data.Map.Strict (Map)
+import           Data.Ratio
+import           Data.Word
+import qualified Network.Socket as Socket
+import           System.Random
+
+import           Cardano.Slotting.Slot (SlotNo)
+import           Ouroboros.Network.PeerSelection.RootPeersDNS (DomainAddress (..))
+import           Ouroboros.Network.Subscription.Ip
+import           Ouroboros.Network.Subscription.Dns
+
+import           Text.Printf
+
+newtype LedgerPeersConsensusInterface m = LedgerPeersConsensusInterface {
+      lpGetPeers :: SlotNo -> STM m (Maybe [(PoolStake, NonEmpty RelayAddress)])
+    }
+
+-- | Trace LedgerPeers events.
+data TraceLedgerPeers =
+      PickedPeer !RelayAddress !AckPoolStake ! PoolStake
+      -- ^ Trace for a peer picked with ackumulated and relative stake of its pool.
+    | PickedPeers !Word16 ![RelayAddress]
+      -- ^ Trace for the number of peers we wanted to pick and the list of peers picked.
+    | FetchingNewLedgerState !Int
+      -- ^ Trace for fetching a new list of peers from the ledger. Int is the number of peers
+      -- returned.
+
+
+instance Show TraceLedgerPeers where
+    show (PickedPeer addr ackStake stake) =
+        printf "PickedPeer %s ack stake %s ( %.04f) relative stake %s ( %.04f )"
+            (show addr)
+            (show $ unAckPoolStake ackStake)
+            (fromRational (unAckPoolStake ackStake) :: Double)
+            (show $ unPoolStake stake)
+            (fromRational (unPoolStake stake) :: Double)
+    show (PickedPeers n peers) =
+        printf "PickedPeers %d %s" n (show peers)
+    show (FetchingNewLedgerState cnt) =
+        printf "Fetching new ledgerstate, %d registered pools"
+            cnt
+
+
+data RelayAddress = RelayAddressDomain DomainAddress
+                  | RelayAddressAddr IP.IP Socket.PortNumber
+                  deriving (Show, Eq, Ord)
+
+-- | The relative stake of the stakepool. A value in the [0, 1] range.
+newtype PoolStake = PoolStake { unPoolStake :: Rational } deriving (Eq, Fractional, Num, Ord, Show)
+
+-- | The relative stake of the stakepool and all preceeding pools. A value in the range [0, 1].
```

Comment:
```suggestion
-- | The relative stake of the stakepool and all preceeding pools. A value in the range [0, 1].
--
```

As above, I would be explicit about relative to what? And I'd explain how this differs from `PoolStake`.